fs-storage/src/folder_storage.rs

```diff
@@ -0,0 +1,573 @@
+pub struct FolderStorage<K, V>
+where
+    K: Ord,
+{
+    /// Label for logging
+    label: String,
+    /// Path to the underlying file where data is persisted
+    path: PathBuf,
+    /// `ram_timestamps` can be used to track the last time a file was modified in memory.
+    /// where the key is the path of the file inside the directory.
+    ram_timestamps: BTreeMap<K, SystemTime>,
+    /// `disk_timestamps` can be used to track the last time a file written or read from disk.
+    /// where the key is the path of the file inside the directory.
+    disk_timestamps: BTreeMap<K, SystemTime>,
+    data: FolderStorageData<K, V>,
+}
+
+/// A struct that represents the data stored in a [`FolderStorage`] instance.
+///
+///
+/// This is the data that is serialized and deserialized to and from disk.
+#[derive(Serialize, Deserialize)]
+pub struct FolderStorageData<K, V>
+where
+    K: Ord,
+{
+    version: i32,
+    entries: BTreeMap<K, V>,
+}
+
+impl<K, V> AsRef<BTreeMap<K, V>> for FolderStorageData<K, V>
+where
+    K: Ord,
+{
+    fn as_ref(&self) -> &BTreeMap<K, V> {
+        &self.entries
+    }
+}
+
+impl<K, V> FolderStorage<K, V>
+where
+    K: Ord
+        + Clone
+        + serde::Serialize
+        + serde::de::DeserializeOwned
+        + std::str::FromStr
+        + std::fmt::Display,
+    V: Clone
+        + serde::Serialize
+        + serde::de::DeserializeOwned
+        + std::str::FromStr
+        + Monoid<V>,
+{
+    /// Create a new folder storage with a diagnostic label and directory path
+    /// The storage will be initialized using the disk data, if the path exists
+    ///
+    /// Note: if the folder storage already exists, the data will be read from the folder
+    /// without overwriting it.
+    pub fn new(label: String, path: &Path) -> Result<Self> {
+        let mut storage = Self {
+            label,
+            path: PathBuf::from(path),
+            ram_timestamps: BTreeMap::new(),
+            disk_timestamps: BTreeMap::new(),
+            data: FolderStorageData {
+                version: STORAGE_VERSION,
+                entries: BTreeMap::new(),
+            },
+        };
+
+        if Path::exists(path) {
+            storage.read_fs()?;
+        }
+
+        Ok(storage)
+    }
+
+    /// Load mapping from folder storage
+    fn load_fs_data(&mut self) -> Result<FolderStorageData<K, V>> {
+        if !self.path.exists() {
+            return Err(ArklibError::Storage(
+                self.label.clone(),
+                "File does not exist".to_owned(),
+            ));
+        }
+
+        if !self.path.is_dir() {
+            return Err(ArklibError::Storage(
+                self.label.clone(),
+                "Path is not a directory".to_owned(),
+            ));
+        }
+
+        let mut data = FolderStorageData {
+            version: STORAGE_VERSION,
+            entries: BTreeMap::new(),
+        };
+
+        self.disk_timestamps.clear();
+        self.ram_timestamps.clear();
```

Comment:
Why do we need to clear the timestamps? Could we add a comment to explain this? Also, adding a test for this would be helpful. All tests pass even without these two lines.

Reply:
I agree timestamps should not be cleared. Any values that need to be changed will be overwritten by the below two lines. Moreover, it stores timestamps of entries that were deleted so clearing it will wipe out that information. 

```rust
                        self.disk_timestamps.insert(key.clone(), modified);
                        self.ram_timestamps.insert(key, modified);
```